include/xtensor-signal/find_peaks.hpp

```diff
@@ -0,0 +1,725 @@
+                    }, _threshold.value());
+
+                peaks = xt::filter(peaks, keep);
+            }
+
+            //check if we want to filter out on distance
+            if(_distance.has_value())
+            {
+                auto keep = detail::select_by_peak_distance(peaks, xt::eval(xt::view(x, xt::keep(peaks))), _distance.value());
+                peaks = xt::filter(peaks, keep);
+            }
+            //check if we want to filter out on prominence
+            if(_prominence.has_value())
+            {
+                auto wlen_safe = detail::arg_wlen_as_expected(_wlen);
+                auto res = detail::peak_prominences(x, peaks, wlen_safe);
+                auto keep = std::visit(detail::Overload{
+                    [&](float arg)
+                    {
+                        return detail::select_by_property(std::get<0>(res), arg);
+                    },
+                    [&](std::pair<float, float> arg)
+                    {
+                        return detail::select_by_property(std::get<0>(res) , arg.first, arg.second);
+                    },
+                    [&](xt::xtensor<float, 1> arg)
+                    {
+                        return detail::select_by_property(std::get<0>(res) , arg);
+                    },
+                    [&](std::pair<xt::xtensor<float, 1>, xt::xtensor<float, 1>> arg)
+                    {
+                       return detail::select_by_property(std::get<0>(res), arg.first, arg.second);
+                    }
+                }, _prominence.value());
+                peaks = xt::filter(peaks, keep);
+            }
+
+            //check if we want to filter out on width
+            if(_width.has_value())
+            {
+                //TODO: should probably capture the case of std vector and convert to array
+                //once we have the prominence we can add it here to avoid recalculating it again
+                auto widths = peak_widths(x, peaks, _rel_height, _wlen);
+                auto keep = std::visit(detail::Overload{
+                    [&](float arg)
+                    {
+                        return detail::select_by_property(std::get<0>(widths), arg);
+                    },
+                    [&](std::pair<float, float> arg)
+                    {
+                        return detail::select_by_property(std::get<0>(widths), arg.first, arg.second);
+                    },
+                    [&](xt::xtensor<float, 1> arg)
+                    {
+                        return detail::select_by_property(std::get<0>(widths), arg);
+                    },
+                    [&](std::pair<xt::xtensor<float, 1>, xt::xtensor<float, 1>> arg)
+                    {
+                       return detail::select_by_property(std::get<0>(widths), arg.first, arg.second);
+                    }
+                }, _width.value());
+                peaks = xt::filter(peaks, keep);
+            }
+            
+            return peaks;
```

Comment:
I think this function could be splitted in 6 parts (one per filter type) as private methods. The `operator()` would then just chain the call to these private methods.